### PR TITLE
coap client: handle allocation errors with sync calls

### PR DESCRIPTION
**Pull request by szczys**
- When creating an event group, check for memory allocation and semaphore creation errors. Free memory as needed and return NULL
- During sync calls, check event group and sempahore for NULL. Free memory as needed and return GOLIOTH_ERROR_MEM_ALLOC

resolves https://github.com/golioth/firmware-issue-tracker/issues/582

### CODE REVIEW

File: src/coap_client.c [hunk 1]
```diff
@@ -421,10 +448,19 @@ static enum golioth_status golioth_coap_client_get_internal(struct golioth_clien
...
+        // Created here, deleted by coap thread (or here if fail to create or enqueue)
         request_msg.request_complete_event = golioth_event_group_create();
         request_msg.request_complete_ack_sem = golioth_sys_sem_create(1, 0);
         request_msg.status = &status;
+
+        if (!request_msg.request_complete_event || !request_msg.request_complete_ack_sem)
+        {
+            GLTH_LOGW(TAG, "Failed to create request_complete members");
+            golioth_event_group_destroy(request_msg.request_complete_event);
+            golioth_sys_sem_destroy(request_msg.request_complete_ack_sem);
```
Comment (sam-golioth) [inline src/coap_client.c:460]: We can't assume that `golioth_sys_sem_destroy()` is tolerant to being called with a semaphore that did not create successfully. For example. FreeRTOS will assert if you attempt to destroy a NULL semaphore, and the ESP-IDF port does not check for NULL.

We should just check each one of these for failure after creating it. There's not an advantage in checking them together other than reducing the lines of code, but that's not an advantage in itself and in this case I believe having separate checks is more readable and maintainable.

**Review by sam-golioth — CHANGES_REQUESTED** [review]

File: src/coap_client.c [hunk 2]
```diff
@@ -162,9 +177,24 @@ static enum golioth_status golioth_coap_client_set_internal(struct golioth_clien
 
     if (is_synchronous)
     {
-        // Created here, deleted by coap thread (or here if fail to enqueue
+        // Created here, deleted by coap thread (or here if fail to create or enqueue)
         request_msg.request_complete_event = golioth_event_group_create();
+        if (!request_msg.request_complete_event)
+        {
+            GLTH_LOGW(TAG, "Failed to create event group");
+            request_msg.status = NULL;
```
Comment (sam-golioth) [inline src/coap_client.c:185]: Same comment as above, this isn't necessary (throughout this, the same is true).

File: src/coap_client.c [hunk 3]
```diff
@@ -162,9 +177,24 @@ static enum golioth_status golioth_coap_client_set_internal(struct golioth_clien
...
+            return GOLIOTH_ERR_MEM_ALLOC;
```
Comment (sam-golioth) [inline src/coap_client.c:186]: We need to free `request_payload` allocated on line 160 above before returning.

**Review by sam-golioth — CHANGES_REQUESTED** [review]

File: src/coap_client.c (hunk 2 above)
Comment (szczys) [reply src/coap_client.c]: Great, I've made a subsequent commit to remove similar code not previously touched by this PR.

File: src/coap_client.c (hunk 3 above)
Comment (szczys) [reply src/coap_client.c:186]: Good catch, thank you!

I made a subsequent commit to change how this is tested a few lines before this.

**Review by sam-golioth — APPROVED** [review]
Looks good!

**Pull request merged**